lib/charms/mongodb/v0/upgrade.py

```diff
@@ -65,30 +82,91 @@ def pre_upgrade_check(self) -> None:
+            raise ClusterNotReadyError(
+                message=default_message, cause="Cluster cannot read/write - please check logs"
+            )
 
         # Future PR - sharding based checks
 
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        if not self.is_cluster_able_to_read_write():
+            raise ClusterNotReadyError(
+                message="post-upgrade check failed and cannot safely upgrade",
+                cause="Cluster cannot read/write",
+            )
+
     @override
     def build_upgrade_stack(self) -> list[int]:
+        """Builds an upgrade stack, specifying the order of nodes to upgrade."""
+        if self.charm.is_role(Config.Role.CONFIG_SERVER):
+            # TODO implement in a future PR a stack for shards and config server
+            pass
+        elif self.charm.is_role(Config.Role.REPLICATION):
+            return self.get_replica_set_upgrade_stack()
+
+    def get_replica_set_upgrade_stack(self) -> list[int]:
         """Builds an upgrade stack, specifying the order of nodes to upgrade.
 
-        TODO Implement in DPE-3940
+        MongoDB Specific: The primary should be upgraded last, so the unit with the primary is
+        put at the very bottom of the stack.
         """
+        upgrade_stack = []
+        units = set([self.charm.unit] + list(self.charm.peers.units))  # type: ignore[reportOptionalMemberAccess]
+        primary_unit_id = None
+        for unit in units:
+            unit_id = int(unit.name.split("/")[-1])
+            if unit.name == self.charm.primary:
+                primary_unit_id = unit_id
+                continue
+
+            upgrade_stack.append(unit_id)
+
+        upgrade_stack.insert(0, primary_unit_id)
+        return upgrade_stack
 
     @override
     def log_rollback_instructions(self) -> None:
-        """Logs the rollback instructions in case of failure to upgrade.
-
-        TODO Implement in DPE-3940
-        """
+        """Logs the rollback instructions in case of failure to upgrade."""
+        logger.critical(ROLLBACK_INSTRUCTIONS)
 
     @override
     def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
-        """Execute a series of upgrade steps.
-
-        TODO Implement in DPE-3940
-        """
+        """Execute a series of upgrade steps."""
+        # TODO: Future PR - check compatibility of new mongod version with current mongos versions
+        self.charm.stop_charm_services()
+
+        try:
+            self.charm.install_snap_packages(packages=Config.SNAP_PACKAGES)
+        except snap.SnapError:
+            logger.error("Unable to install Snap")
+            self.set_unit_failed()
+            return
+
+        logger.info(f"{self.charm.unit.name} upgrading service...")
+        self.charm.restart_charm_services()
+
+        try:
+            logger.debug("Running post-upgrade check...")
+            for attempt in Retrying(
```

Comment:
**nitpick:** could this retrying block decorate `post_upgrade_check`? Or is this a preference thing?

Reply:
Its a actually an optimisation thing, restartiing mongod can be very costly and so its better to wrap just `post_upgrade_check`.

Great question!

Reply:
Ah smart 🧠